docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+  BasePath string
+
+  // Path to the swagger spec.
+  SwaggerSpecPath string
+}
+
+type ThirdPartyAPIServerStatus struct {
+  Conditions []ThirdPartyAPIServerCondition
+}
+
+type ThirdPartyAPIServerConditionType string
+const (
+        ThirdPartyAPIServerValid ThirdPartyAPIServerConditionType = "Valid"
+)
+
+type ThirdPartyAPIServerCondition struct {
+  Type ThirdPartyAPIServerConditionType
+  // Status of the condition, one of True, False or Unknown.
+  Status ConditionStatus
+  Reason string
+  Message string
+}
+```
+
+There are a few requirements that the third party API servers should meet
+(for ex: name of resources provided by each server should be unique).
+ThirdPartyAPIServerController will look for new servers and will validate them.
+It will set the valid condition in the status accordingly.
+Note that the server is validated only once so it is possible to change it after
+the validation is done. We expect that the server will be deleted and added
+again, whenever its functionality is changed.
+
+### kubectl
+
+Users will be able to use resources provided by third party servers seamlessly
+as if they were provided by the core server.
+Kubectl today uses a RESTMapper to determine the list of valid resources and
+operations. We will switch it to use the discovery API so that we can add and remove resources
+at runtime. It will first get the list of third party servers from the core server
+and then use their discovery APIs to get the list of resources and operations
+supported by each one of them.
+
+### Events
+
+Core server will store all events. Third party servers will POST their events to
+the core server.
+As a result, users will get all events in the cluster, when they run `kubectl
+get events` or GET `kubernetes-master/api/v1/events`.
+
+### Auth
+
+Each API server will be resposible for authentication and authorisation for
+their own resources. Right now, the code will be duplicated in all servers.
+In future, we can provide a service which all of them can call.
+
+## Caveats
+
+* Resources can have same uid: kubectl users, for example, can get 2 resources
+  with the same uid. Note that the resources will be of different kind from different
+  API servers and should not cause any problem, but it is an important change to
+  call out.
+
+## Alternatives
+
+There were other alternatives that we had discussed, but decided against them
+
+* Core APIServer taking care of auth for everyone: Instead of every APIServer
+  doing the auth, another alternative was to let the core APIServer do the
+  auth. But this required all client requests to go through the core APIServer,
+  which was an unnecessary burden on it.
+* Adding a proxy in front of all API servers that directs request to the
+  appropriate server: Instead of letting clients talk to each server directly,
+  we can add a proxy layer which the clients always talk to. We decided not
+  to do it as letting clients directly talk to each server is fine.
```

Comment:
We can always implement that pattern using this proposal later, simply by exposing an authorizing proxy at /apis/ and returning that URL to the client.  In fact, I think we should move this into "future steps"